### ISSUE TITLE
Non deterministic behaviour when reading variables with lifetimes

### ISSUE THREAD

**Issue by Joshix-1** [vivaansinghvi07/dreamberd-interpreter#46]
I have the following script. It's supposed to count from -1 to 2220 while waiting 1 second between every print.
```db
var const i = -1!

when (i + 1 < 2222) {
   const const time<1s> = waiting!

   var const j = one!
   when(j) {
      if (time == waiting) {
         j = j + one!
      }
      if (time ;= waiting) {
         print(i)!
         i = i + one!
      }
   }
}
```

Sometimes it prints:
```
Warning: Public global variable `pivo` access failed.
-1
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
dreamberd.base.NonFormattedError: Variable is undefined.

During handling of the above exception, another exception occurred:

dreamberd.base.InterpretationError: __unnamed_file__, line 8

        if (time == waiting) {

Variable is undefined.
```
Sometimes only -1:
```
Warning: Public global variable `pivo` access failed.
-1
dreamberd.base.NonFormattedError: Variable is undefined.

During handling of the above exception, another exception occurred:

dreamberd.base.InterpretationError: __unnamed_file__, line 8

        if (time == waiting) {

Variable is undefined.
```

There's probably some kind of race-condition when trying to remove the variables from the thread. I guess a lock somewhere could fix it.

**Comment by Joshix-1** [comment]
The following patch makes the issue less likely to occur:
```diff
diff --git a/dreamberd/builtin.py b/dreamberd/builtin.py
index f1bc374..6fa1cc8 100644
--- a/dreamberd/builtin.py
+++ b/dreamberd/builtin.py
@@ -376,8 +376,8 @@ class Variable:
     def value(self) -> DreamberdValue:
         if self.lifetimes:
             return self.lifetimes[0].value
-        raise NonFormattedError("Variable is undefined.")
-    
+        return DreamberdString(self.name)
+
 def all_function_keywords() -> list[str]:
 
     # this code boutta be crazy
 ````

**Comment by Joshix-1** [comment]
But with that patch i get after 78 the error `RecursionError: maximum recursion depth exceeded`

**Comment by vivaansinghvi07** [comment]
Thank you for your contributions! I think if that patch seems to work I would just increase the maximum recursion limit in the main `__init__.py` file, as during creation I made the bad design decision of checking if when statements are activated within variable assignment, which ends up recursing into `assign -> run when statement -> assign -> run when statement -> ...`

**Comment by vivaansinghvi07** [comment]
Actually, nevermind -- I forgot I had already done that; I think the constant updating of `j` would just have to be changed to be less often or something as each update adds a few more stack frames.

**Comment by Joshix-1** [comment]
After thinking about the patch i think it's bad. It probably should also check if the name is numeric and then return a numeric value.

the whole reason for the j updates is, that there is no gold way to sleep. I tried to just listen with when on the end of lifetime, but that didn't work. (As described in the Readme)

i've locally increased the stack limit even further and that resulted in really high ram usage (over 40GB). So that's not really practical

it would probably best to not do recursive calls

**Comment by vivaansinghvi07** [comment]
Yeah, changing the inner workings like that definitely would take a while though -- I think for the time being we could add a sleep function into the language

**Comment by vivaansinghvi07** [comment]
I just pushed a commit that adds a sleep function so it should make your use case a lot easier to implement

**Comment by Joshix-1** [comment]
A sleep function is kind of boring. The original spec didn't mention it

**Comment by not-nullptr** [comment]
This is intended behaviour - do not question the Perfect Language. All behaviour is intended behaviour.
